### ISSUE TITLE
useIsMutating hangs application

### ISSUE THREAD

**Issue by mrlika** [TanStack/query#6717]
### Describe the bug

A combination of `useIsMutating` and `useMutation` hooks causes hang.

We found a reason why our production apps hang with `Maximum update depth exceeded.` React error after updating to the latest `@tanstack/react-query`

### Your minimal, reproducible example

https://stackblitz.com/edit/stackblitz-starters-q54tlx?description=React%20%20%20TypeScript%20starter%20project&file=src%2FApp.tsx&title=React%20Starter

- Replacing to `useIsMutating({ mutationKey: ['MUTATION_2'] })` or removing the line fixes the app
- Downgrading to `@tanstack/react-query@5.17.9` fixes the app
- Making mutate calls not simultaneous fixes the app
- Reverting https://github.com/TanStack/query/pull/6693 fixes the app

### Steps to reproduce

Click `TEST` button.

### Expected behavior

No errors in the console.

### How often does this bug happen?

Every time

### Screenshots or Videos

![image](https://github.com/TanStack/query/assets/1469266/3647b9f4-f3ac-416c-a4cc-542a0f82f965)

### Platform

Platform or browser doesn't matter.

### Tanstack Query adapter

react-query

### TanStack Query version

5.17.12

### TypeScript version

v5

**Comment by mrlika** [comment]
Reverting https://github.com/TanStack/query/pull/6693/files fixes the issue

**Comment by TkDodo** [comment]
thanks for reporting. I narrowed it down to this minimal test case for `replaceEqualDeep`:

```
it('should be able to share values that contain undefined', () => {
  const current = [
    {
      data: undefined,
      foo: true,
    },
  ]

  const next = replaceEqualDeep(current, [
    {
      data: undefined,
      foo: true,
    },
  ])

  expect(current).toBe(next)
})
```

fails with:

```
AssertionError: expected [ { data: undefined, foo: true } ] to be [ { data: undefined, foo: true } ] // Object.is equality

If it should pass with deep equality, replace "toBe" with "toStrictEqual"

Expected: [ { data: undefined, foo: true } ]
Received: serializes to the same string
```

@schiller-manuel FYI

since this traces back to https://github.com/TanStack/router/pull/903, I'm also pulling in the original author of that PR, @kahosan

Any help is appreciated; if we can't find a quick way to fix this, I'll revert #6693
